support-frontend/assets/helpers/subscriptionsForms/formValidation.ts

```diff
@@ -21,94 +21,118 @@ import { getFulfilmentOption } from 'helpers/subscriptionsForms/subscriptionChec
-	dispatchErrors(dispatch, allErrors);
+	dispatchAllErrors(dispatch, allErrors);
 	return allErrors.length === 0;
 }
 
-const checkoutFormIsValid = (state: SubscriptionsState) => {
+export function checkoutFormIsValid(state: SubscriptionsState): boolean {
 	if (state.page.checkoutForm.product.productType === DigitalPack) {
 		return checkoutValidation(state).length === 0;
 	}
 	return withDeliveryValidation(state).length === 0;
-};
+}
 
-const withDeliveryFormIsValid = (state: SubscriptionsState) =>
-	withDeliveryValidation(state).length === 0;
+export function withDeliveryFormIsValid(state: SubscriptionsState): boolean {
+	return withDeliveryValidation(state).length === 0;
+}
 
-export {
-	validateWithDeliveryForm,
-	validateCheckoutForm,
-	checkoutFormIsValid,
-	withDeliveryFormIsValid,
-};
+// ---- Helpers ---- //
+
+function checkoutValidation(state: SubscriptionsState): AnyErrorType[] {
+	const errors: AnyErrorType[] = [];
+	const pushIfErrors = getPushIfErrors(errors);
+
+	const checkoutErrors = applyCheckoutRules(getFormFields(state));
+	pushIfErrors({
+		errors: checkoutErrors,
+		dispatchErrors: (dispatch) => dispatch(setFormErrors(checkoutErrors)),
+	});
+
+	const billingAddressErrors = applyBillingAddressRules(
+		state.page.checkoutForm.billingAddress.fields,
+	);
+	pushIfErrors({
+		errors: billingAddressErrors,
+		dispatchErrors: (dispatch) =>
+			dispatch(setBillingAddressFormErrors(billingAddressErrors)),
+	});
+
+	return errors;
+}
+
+function withDeliveryValidation(state: SubscriptionsState): AnyErrorType[] {
+	const errors: AnyErrorType[] = [];
+	const pushIfErrors = getPushIfErrors(errors);
+
+	const formFields = getFormFields(state);
+
+	const deliveryErrrors = applyDeliveryRules(formFields);
+	pushIfErrors({
+		errors: deliveryErrrors,
+		dispatchErrors: (dispatch) => dispatch(setFormErrors(deliveryErrrors)),
+	});
+
+	const deliveryAddressErrors = applyDeliveryAddressRules(
+		getFulfilmentOption(state),
+		state.page.checkoutForm.deliveryAddress.fields,
+	);
+	pushIfErrors({
+		errors: deliveryAddressErrors,
+		dispatchErrors: (dispatch) =>
+			dispatch(setDeliveryAddressFormErrors(deliveryAddressErrors)),
+	});
+
+	if (shouldValidateBillingAddress(formFields)) {
+		const billingAddressErrors = applyBillingAddressRules(
+			state.page.checkoutForm.billingAddress.fields,
+		);
+		pushIfErrors({
+			errors: billingAddressErrors,
+			dispatchErrors: (dispatch) =>
+				dispatch(setBillingAddressFormErrors(billingAddressErrors)),
+		});
+	}
+
+	return errors;
+}
+
+function shouldValidateBillingAddress(fields: FormFields) {
+	return !fields.billingAddressIsSame;
+}
+
+function dispatchAllErrors(dispatch: Dispatch, allErrors: AnyErrorType[]) {
+	allErrors.forEach((e) => e.dispatchErrors(dispatch));
+}
+
+function getPushIfErrors(errors: AnyErrorType[]) {
+	function pushIfErrors(error: AnyErrorType) {
+		if (error.errors.length > 0) {
+			errors.push(error);
+		}
+	}
+	return pushIfErrors;
+}
```

Comment:
I factored this out because there were lots of repeated blocks checking `length > 0` and then pushing to the errors array. However, I can't quite decide if it's better or it just adds too much complexity

Reply:
I really like the idea of extracting this re-used condition, but I think it might be a bit of a footgun to have this be causing a side effect- it's not immediately obvious when you're looking at the validation functions that `pushIfErrors` is changing the `errors` array.

It might be simpler to have the condition inside a function that simply returns an array, which may or may not be empty, eg:

```typescript
function getErrorList(error: AnyErrorType) {
	if (error.errors.length > 0) {
		return [error]
	}
	return [];
}

function checkoutValidation() {
	const checkoutErrors = getErrorList({
		errors: applyCheckoutRules(getFormFields(state)),
		dispatchErrors: (dispatch) => dispatch(setFormErrors(checkoutErrors)),
	});
	
	const billingAddressErrors = getErrorList({
		errors: applyBillingAddressRules(
			state.page.checkoutForm.billingAddress.fields,
		),
		dispatchErrors: (dispatch) => dispatch(setFormErrors(checkoutErrors)),
	});

	return [...checkoutErrors, ...billingAddressErrors];
```

Reply:
Yeah agreed the side effects make it really unclear. 

So just a small point there, the `dispatchErrors` function needs a reference to the `errors` array, so the code would have to look like:

```ts
function checkoutValidation(state: SubscriptionsState): AnyErrorType[] {
	const checkoutErrors = applyCheckoutRules(getFormFields(state));
	const checkoutErrorsList = getErrorList({
		errors: checkoutErrors,
		dispatchErrors: (dispatch) => dispatch(setFormErrors(checkoutErrors)),
	});

	const billingAddressErrors = applyBillingAddressRules(
		state.page.checkoutForm.billingAddress.fields,
	);
	const billingAddressErrorsList = getErrorList({
		errors: billingAddressErrors,
		dispatchErrors: (dispatch) =>
			dispatch(setBillingAddressFormErrors(billingAddressErrors)),
	});

	return [...checkoutErrorsList, ...billingAddressErrorsList];
}
```

Reply:
I think this is definitely an improvement. It still feels like were having to work hard to get the code to do what we want here. Perhaps there's a bigger refactor that could really clean things up.

Reply:
Okay, I've pushed that up!